src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocPackageCheck.java

```diff
@@ -0,0 +1,77 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2019 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.FileContents;
+import com.puppycrawl.tools.checkstyle.api.TextBlock;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * <p>
+ * Checks for missing Javadoc comments in package-info.java files.
+ * </p>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="MissingJavadocPackage"/&gt;
+ * </pre>
+ *
+ * @since 8.22
+ */
+@StatelessCheck
+public class MissingJavadocPackageCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_PKG_JAVADOC_MISSING = "package.javadoc.missing";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.PACKAGE_DEF};
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final FileContents contents = getFileContents();
```

Comment:
Can we use https://checkstyle.org/writingchecks.html#Comment_Tokens_in_AST ?
 (you can find examples of Checks in our codebase), to do all by AST traversing.

Reply:
I tried, but there is a bug(?) in parser. If there is annotation between javadoc and package definition, it sees only package def.
```
/**
* javadoc
*/
@Annotation 
package com.checkstyle.pack; <-- parser recognizes only this
```

Therefore I used filecontents way to detect violation.

Reply:
Upd. I checked what tree cli shows.
For package-info file with annotation and comment it looks like:
```
 java -jar checkstyle-8.20-all.jar -T package-info.java
PACKAGE_DEF -> package [5:0]
|--ANNOTATIONS -> ANNOTATIONS [4:0]
|   `--ANNOTATION -> ANNOTATION [4:0]
|       |--BLOCK_COMMENT_BEGIN -> /* [1:0]
|       |   |--COMMENT_CONTENT -> *\r\n* Important javadoc\r\n [1:2]
|       |   `--BLOCK_COMMENT_END -> */ [3:0]
|       |--AT -> @ [4:0]
|       `--IDENT -> Deprecated [4:1]
|--DOT -> . [5:17]
|   |--DOT -> . [5:11]
|   |   |--IDENT -> com [5:8]
|   |   `--IDENT -> strkk [5:12]
|   `--IDENT -> checkstyle [5:18]
`--SEMI -> ; [5:28]
```
So annotation and comment are children of package_def.
As for me, it is a bit counterintuitive.
I will try to refactor check with only ast.

Reply:
yes, this is new bug in our parser, as we missed a point that javadoc could be on package.
https://github.com/checkstyle/checkstyle/issues/4392

So it is bug and by design.
But still it is not a reason to avoid AST, please implement logic to expect commit in two places. Ones we resolve issue in parser to place comments always at one place for "package" we will update a Check.